### ISSUE TITLE
数据中各个参数的含义

### ISSUE THREAD

**Issue by du-67** [ls1rius/WSI_FiVE#17]
您好！请问数据的 patch_maxr 、patch_maxc以及patch_pub_cnt、patch_inds这些参数分别代表什么呢？您是如何得到的呢？
非常期待您的回复！谢谢！

**Comment by ls1rius** [comment]
patch相对于wsi原图的切片位置信息，我做别的实验时遗留的数据，在这个项目里用不到。

**Comment by du-67** [comment]
好的，请问patch_maxr 、patch_maxc以及patch_pub_cnt、patch_inds这些参数在这个项目中不需要吗？但我看生成图像特征时的函数中用到了patch_pub_cnt、patch_inds这两个参数：  self.mit = PatchFusionTransformer(T, embed_dim=embed_dim, layers_sa=1, layers_ca=1,) 
 class MultiPatchIntegrationTransformer(nn.Module):
    def __init__(self, T, embed_dim=512, layers=1,):
        super().__init__()
        self.T = T
        transformer_heads = embed_dim // 64
        self.positional_embedding = nn.Parameter(torch.empty(1, T, embed_dim))
        trunc_normal_(self.positional_embedding, std=0.02)
        self.resblocks = nn.Sequential(*[ResidualAttentionBlock(d_model=embed_dim, n_head=transformer_heads) for _ in range(layers)])
        self.pad_embed = nn.Embedding(1, embed_dim)

        self.apply(self._init_weights)

    def _init_weights(self, m):
        if isinstance(m, (nn.Linear,)):
            trunc_normal_(m.weight, std=0.02)
            if m.bias is not None:
                nn.init.zeros_(m.bias)
        elif isinstance(m, nn.LayerNorm):
            nn.init.zeros_(m.bias)
            nn.init.ones_(m.weight)

    def forward(self, x, prompt, patch_info):
        # unpack patch info data
        sample_range = patch_info['sample_range']
        posis_rlt, posis_enc = get_pos_embed(patch_info, x.shape[-1])
        # posx_rlt, posy_rlt, posx_enc, posy_enc = get_pos_embed(patch_info, x.shape[-1])

        # pad x data
        x_new = self.pad_embed(torch.zeros(x.shape[0], x.shape[1], dtype=torch.long, device=x.device))
        for i in range(len(sample_range)):
            x_new[i, :sample_range[i]] = x[i, :sample_range[i]]
        x = x_new

        # attn process
        ori_x = x
        # if prompt is None:
        #     x = x + posx_enc.to(x.device) + posy_enc.to(x.device)
        x = x + posis_enc.to(x.device)
        # x = x + posx_enc.to(x.device) + posy_enc.to(x.device)
        # x = x + self.positional_embedding
        x = x.permute(1, 0, 2)
        x, _, _ = self.resblocks([x, prompt, sample_range])
        x = x.permute(1, 0, 2)

        if prompt is None:
            x = x.type(ori_x.dtype) + ori_x
        return x

class PatchFusionTransformer(nn.Module):
    def __init__(self, T, embed_dim=512, layers_sa=1, layers_ca=1):
        super().__init__()
        self.self_attn = MultiPatchIntegrationTransformer(T, embed_dim, layers_sa)
        self.cross_attn = MultiPatchIntegrationTransformer(T, embed_dim, layers_ca)
        self.bn1 = nn.LayerNorm(embed_dim)
        self.bn2 = nn.LayerNorm(embed_dim)
        # self.project_for_cls = nn.Sequential(OrderedDict([
        #     ('fc1', nn.Linear(embed_dim, embed_dim//16)),
        #     ('relu1', nn.LeakyReLU()),
        #     ('fc2', nn.Linear(embed_dim/16, embed_dim)),
        # ]))
        # self.project_for_cls = nn.Identity()
        self.project_for_cls = nn.Linear(embed_dim * 2, embed_dim)
        # self.project_for_cls = nn.Linear(embed_dim, embed_dim)

    def forward(self, x, prompt, patch_info):
        sample_range = patch_info['sample_range']

        x = self.self_attn(x, None, patch_info)
        x_new = []
        for i in range(len(sample_range)):
            x_new.append(x[i, :sample_range[i]].sum(dim=0) / sample_range[i])
        x_new = torch.stack(x_new)

        x = self.cross_attn(x, prompt, patch_info)
        x = x.mean(dim=1, keepdim=False)

        # whether to fusion
        x = self.project_for_cls(torch.cat([x, x_new], dim=1))
        # x = self.project_for_cls(x + x_new)
        # x = self.project_for_cls(self.bn1(x) + self.bn2(x_new))
        return x

def get_pos_embed(patch_info, d_model):
    posis = patch_info['patch_inds']
    patch_pub_cnt = patch_info['patch_pub_cnt']
    ###
    posis_rlt = posis / patch_pub_cnt.reshape(-1, 1)
    posis_enc = torch.zeros((posis_rlt.shape[0], posis_rlt.shape[1], d_model))

    for i in range(d_model):
        if i % 2 == 0:
            posis_enc[..., i] = torch.sin(posis_rlt / (10000 ** (2 * i / d_model)))
        else:
            posis_enc[..., i] = torch.cos(posis_rlt / (10000 ** ((2 * i - 1) / d_model)))

    return posis_rlt, posis_enc

**Comment by ls1rius** [comment]
抱歉，我check了下。
首先这些标记表示patch相对于wsi原图的切片位置信息，分别表示横纵坐标以及相对次序。
然后代码中这一段，最初是有尝试过按照patch在原wsi上的位置信息对其进行位置编码，后面也尝试了直接按顺序进行编码，发现差别不大，所以这里就采用了简单的位置编码方式。

**Comment by du-67** [comment]
感谢您的回复。我目前在用自己数据集进行项目复现，但没有patch_maxr 、patch_maxc以及patch_pub_cnt、patch_inds这些参数，请问这些参数您是如何得到的呢？或者如果不使用这些参数能进行实验吗？

**Comment by ls1rius** [comment]
分别是WSI切片成patch后的最大行最大列，patch切片计数，当前计数下标。
可以用传统方式生成位置编码对实验结果影响不大。

**Comment by du-67** [comment]
明白了，非常感谢您的回复！
